Xamarin.Forms.Platform.WPF/Renderers/MediaElementRenderer.cs

```diff
@@ -0,0 +1,299 @@
+					Control.BufferingEnded -= ControlBufferingEnded;
+					Control.MediaOpened -= ControlMediaOpened;
+					Control.MediaEnded -= ControlMediaEnded;
+					Control.MediaFailed -= ControlMediaFailed;
+				}
+			}
+
+			if (e.NewElement != null)
+			{
+				SetNativeControl(new System.Windows.Controls.MediaElement());
+				Control.HorizontalAlignment = HorizontalAlignment.Stretch;
+				Control.VerticalAlignment = VerticalAlignment.Stretch;
+
+				Control.LoadedBehavior = MediaState.Manual;
+				Control.UnloadedBehavior = MediaState.Close;
+				Control.Stretch = Element.Aspect.ToStretch();
+
+				Control.BufferingStarted += ControlBufferingStarted;
+				Control.BufferingEnded += ControlBufferingEnded;
+				Control.MediaOpened += ControlMediaOpened;
+				Control.MediaEnded += ControlMediaEnded;
+				Control.MediaFailed += ControlMediaFailed;
+
+				Element.SeekRequested += ElementSeekRequested;
+				Element.StateRequested += ElementStateRequested;
+				Element.PositionRequested += ElementPositionRequested;
+				UpdateSource();
+			}
+		}
+
+		void ElementPositionRequested(object sender, EventArgs e)
+		{
+			if (!(Control is null))
+			{
+				Controller.Position = Control.Position;
+			}
+		}
+		
+		void ElementStateRequested(object sender, StateRequested e)
+		{
+			_requestedState = e.State;
+
+			switch (e.State)
+			{
+				case MediaElementState.Playing:
+					if (Element.KeepScreenOn)
+					{
+						DisplayRequestActive();
+					}
+
+					Control.Play();
+					Controller.CurrentState = _requestedState;
+					break;
+
+				case MediaElementState.Paused:
+					if (Control.CanPause)
+					{
+						if (Element.KeepScreenOn)
+						{
+							DisplayRequestRelease();
+						}
+
+						Control.Pause();
+						Controller.CurrentState = _requestedState;
+					}
+					break;
+
+				case MediaElementState.Stopped:
+					if (Element.KeepScreenOn)
+					{
+						DisplayRequestRelease();
+					}
+
+					Control.Stop();
+					Controller.CurrentState = _requestedState;
+					break;
+			}
+
+			Controller.Position = Control.Position;
+		}
+
+		void ElementSeekRequested(object sender, SeekRequested e)
+		{
+			Control.Position = e.Position;
+			Controller.Position = Control.Position;
+		}
+
+		void UpdateSource()
+		{
+			if (Element.Source is null)
+				return;
+
+			if (Control.Clock != null)
+				Control.Clock = null;
+
+			if (Element.Source is UriMediaSource uriSource)
+			{
+				if (uriSource.Uri.Scheme == "ms-appx")
+				{
+					Control.Source = new Uri(uriSource.Uri.ToString().Replace("ms-appx://", "pack://application:,,,"));
+				}
+				else if (uriSource.Uri.Scheme == "ms-appdata")
+				{
+					string filePath = string.Empty;
+
+					if (uriSource.Uri.LocalPath.StartsWith("/local"))
+					{
+						// WPF doesn't have the concept of an app package local folder so using My Documents as a placeholder
+						filePath = System.IO.Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments), uriSource.Uri.LocalPath.Substring(7));
```

Comment:
Maybe this should be in the `ApplicationData` path rather? https://docs.microsoft.com/en-us/dotnet/api/system.environment.specialfolder?view=netframework-4.7.2

Reply:
I think we'd have to create a folder beneath this with the app name, will investigate further

Reply:
This returns (for example) C:\Users\username\AppData\Roaming so we'd have to append either an appname or a company/appname. This has an impact on how the rest of the application data is stored and I'm not sure which convention to settle on?